lib/model/topics.dart

```diff
@@ -0,0 +1,170 @@
+
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+
+import '../api/model/events.dart';
+import '../api/model/model.dart';
+import '../api/route/channels.dart';
+import 'channel.dart';
+import 'store.dart';
+
+// similar to _apiSendMessage in lib/model/message.dart
+final _apiGetChannelTopics = getChannelTopics;
+
+/// The view-model for tracking channel topics.
+///
+/// Use [getChannelTopics] to first fetch the topics for a channel from the
+/// server and then use [channelTopics] to get those topics, affected
+/// by the relevant events.
+class Topics extends PerAccountStoreBase with ChangeNotifier {
+  Topics({required super.core});
+
+  /// Maps indexed by channel IDs, of the known latest message IDs in each topic.
+  ///
+  /// For example: `_latestMessageIdsByChannelTopic[channel.streamId][topic] = maxId`
+  ///
+  /// Occasionally, the latest message ID of a topic will refer to a message
+  /// that doesn't exist or is no longer in the topic.
+  /// This happens when the topic's latest message is deleted or moved
+  /// and we don't have enough information to replace it accurately.
+  /// (We don't keep a snapshot of all messages.)
+  // TODO(#2004): handle more cases where this can change
+  final Map<int, TopicKeyedMap<int>> _latestMessageIdsByChannelTopic = {};
+
+  /// Map of topics per channel, sorted by latest message IDs descending.
+  ///
+  /// Derived from [_latestMessageIdsByChannelTopic];
+  /// used for optimized topics sorting.
+  ///
+  /// A channel entry should be discarded when there is a change to
+  /// the ordering of its topics or the channel is no longer present
+  /// in [_latestMessageIdsByChannelTopic], allowing [channelTopics]
+  /// to recalculate the sorted topics on demand.
+  // TODO(#2004): handle more cases where this can change
+  final Map<int, List<GetChannelTopicsEntry>> _sortedTopicsByChannel = {};
+
+  final Map<int, Future<List<GetChannelTopicsEntry>>> _channelTopicsFetching = {};
+
+  /// Get the topics in a channel; typically making a request to the server.
+  ///
+  /// This will avoid making redundant requests for the same data,
+  /// even if this method is called many times. The futures returned from each
+  /// call for the same channel will all complete once the data is ready.
+  ///
+  /// Once fetched, the data will be updated by events;
+  /// use [channelTopics] to consume the data.
+  Future<List<GetChannelTopicsEntry>> getChannelTopics(int channelId) async {
+    List<GetChannelTopicsEntry>? topics = channelTopics(channelId);
+    if (topics != null) return topics;
+
+    // See if another call has already started fetching topics for this channel.
+    Future<List<GetChannelTopicsEntry>>? future = _channelTopicsFetching[channelId];
+    if (future != null) return future;
+
+    future = _getChannelTopics(channelId);
+    _channelTopicsFetching[channelId] = future;
+
+    try {
+      topics = await future;
+      notifyListeners();
```

Comment:
nit: I think it's a bit clearer for the reader to verify the logic here if this `notifyListeners` happens next to the mutations that make it necessary, i.e. in `_getChannelTopics` where we update `_latestMessageIdsByChannelTopic`.